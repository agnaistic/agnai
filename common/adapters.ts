import { AppSchema } from './types/schema'

export type AIAdapter = (typeof AI_ADAPTERS)[number]
export type ChatAdapter = (typeof CHAT_ADAPTERS)[number]
export type PersonaFormat = (typeof PERSONA_FORMATS)[number]

export type AdapterSetting = {
  /** The name of the field within the settings object */
  field: string

  /** The name as it appears in the Settings UI */
  label: string

  /** Any additional information about the setting (for UI only) */
  helperText?: string

  /** If this is a secret that should be encrypted */
  secret: boolean

  /** If the field should be hidden from the UI */
  hidden?: boolean

  setting: SettingType
  preset?: boolean
}

type SettingType =
  | { type: 'list'; options: Array<{ label: string; value: string }> }
  | { type: 'text'; placeholder?: string }
  | { type: 'boolean' }

export type AdapterOptions = {
  /** Name of the adapter that will be displayed in the UI */
  label: string
  settings: AdapterSetting[]
  options: Array<keyof PresetAISettings>
}

export const PERSONA_FORMATS = ['boostyle', 'wpp', 'sbf', 'text'] as const

export const PERSONA_LABELS: { [key in PersonaFormat]: string } = {
  boostyle: 'Boostyle',
  wpp: 'W++',
  sbf: 'SBF',
  text: 'Plain Text',
}

export const AI_ADAPTERS = [
  'kobold',
  'novel',
  'ooba',
  'horde',
  'openai',
  'scale',
  'claude',
  'goose',
  'replicate',
  'openrouter',
  'mancer',
] as const
export const CHAT_ADAPTERS = ['default', ...AI_ADAPTERS] as const

export const MULTI_TENANT_ADAPTERS = ['novel', 'chai', 'kobold'] as const

export type NovelModel = keyof typeof NOVEL_MODELS

export type OpenAIModel = (typeof OPENAI_MODELS)[keyof typeof OPENAI_MODELS]

export const GOOSE_ENGINES = {
  'cassandra-lit-2-8b': 'Cassandra 2.8B',
  'cassandra-lit-6-9b': 'Cassandra 6.9B',
  'convo-6b': 'Convo 6B',
  'gpt-neo-20b': 'GPT Neo 20B',
  'gpt-j-6b': 'GPT-J 6B',
  'gpt-neo-2-7b': 'GPT-Neo 2.7B',
  'gpt-neo-1-3b': 'GPT-Neo 1.3B',
  'gpt-neo-125m': 'GPT-Neo 125M',
  'fairseq-13b': 'Fairseq 13B',
  'fairseq-6-7b': 'Fairseq 6.7B',
  'fairseq-2-7b': 'Fairseq 2.7B',
  'fairseq-1-3b': 'Fairseq 1.3B',
  'fairseq-125m': 'Fairseq 125M',
}

export const OPENAI_MODELS = {
  DaVinci: 'text-davinci-003',
  Turbo: 'gpt-3.5-turbo',
  Turbo0301: 'gpt-3.5-turbo-0301',
  Turbo0613: 'gpt-3.5-turbo-0613',
  Turbo_16k: 'gpt-3.5-turbo-16k',
  GPT4: 'gpt-4',
  GPT4_0314: 'gpt-4-0314',
  GPT4_0613: 'gpt-4-0613',
  GPT4_32k: 'gpt-4-32k',
  GPT4_32k_0314: 'gpt-4-32k-0314',
  GPT4_32k_0613: 'gpt-4-32k-0613',
} as const

export const OPENAI_CHAT_MODELS: Record<string, boolean> = {
  [OPENAI_MODELS.Turbo]: true,
  [OPENAI_MODELS.Turbo0301]: true,
  [OPENAI_MODELS.Turbo0613]: true,
  [OPENAI_MODELS.Turbo_16k]: true,
  [OPENAI_MODELS.GPT4]: true,
  [OPENAI_MODELS.GPT4_0314]: true,
  [OPENAI_MODELS.GPT4_0613]: true,
  [OPENAI_MODELS.GPT4_32k]: true,
  [OPENAI_MODELS.GPT4_32k_0314]: true,
  [OPENAI_MODELS.GPT4_32k_0613]: true,
}

/** Note: claude-v1 and claude-instant-v1 not included as they may point
 * to different models in the future. New models may be less appropriate
 * for roleplaying so they should be updated to manually
 * <https://console.anthropic.com/docs/api/reference#-v1-complete>
 */
export const CLAUDE_MODELS = {
  ClaudeV1: 'claude-v1',
  ClaudeV2: 'claude-2',
  ClaudeV1_100k: 'claude-v1-100k',
  ClaudeV1_0: 'claude-v1.0',
  ClaudeV1_2: 'claude-v1.2',
  ClaudeV1_3: 'claude-v1.3',
  ClaudeV1_3_100k: 'claude-v1.3-100k',
  ClaudeInstantV1: 'claude-instant-v1',
  ClaudeInstantV1_100k: 'claude-instant-v1-100k',
  ClaudeInstantV1_0: 'claude-instant-v1.0',
  ClaudeInstantV1_1: 'claude-instant-v1.1',
  ClaudeInstantV1_1_100k: 'claude-instant-v1.1-100k',
} as const

export const NOVEL_MODELS = {
  euterpe: 'euterpe-v2',
  krake: 'krake-v2',
  clio_v1: 'clio-v1',
  kayra_v1: 'kayra-v1',
} satisfies { [key: string]: string }

export const REPLICATE_MODEL_TYPES = {
  'Auto-detect': '',
  LLaMa: 'llama',
  StableLM: 'stablelm',
  'Open Assistant': 'openassistant',
} as const

export type OpenRouterModel = {
  id: string
  pricing: { prompt: string; completion: string }
  context_length: number
}

export type HordeModel = {
  name: string
  count: number
  performance: number
  queued: number
  eta: number
  type?: string
}

export type HordeWorker = {
  id: string
  name: string
  type: 'text' | 'image'
  online: number
  max_length: number
  max_context_length: number

  requests_fulfilled: number
  kudos_rewards: number
  kudos_details: {
    generated: number
    uptime?: number
  }
  performance: string
  threads: number
  uptime: number
  maintenance_mode: boolean
  nsfw: boolean
  trusted: boolean
  flagged: false
  uncompleted_jobs: number
  models: string[]
  team: {
    name?: string
    id?: string
  }
  bridge_agent: string
}

export const ADAPTER_LABELS: { [key in AIAdapter]: string } = {
  horde: 'Horde',
  kobold: 'Kobold / 3rd Party',
  novel: 'NovelAI',
  ooba: 'Textgen',
  openai: 'OpenAI',
  scale: 'Scale',
  claude: 'Claude',
  goose: 'Goose AI',
  replicate: 'Replicate',
  openrouter: 'OpenRouter',
  mancer: 'Mancer',
}

export const INSTRUCT_SERVICES: { [key in AIAdapter]?: boolean } = {
  openai: true,
  openrouter: true,
  claude: true,
  scale: true,
  novel: true,
}

export type PresetAISettings = Omit<
  AppSchema.GenSettings,
  | 'name'
  | 'service'
  | 'images'
  | 'memoryDepth'
  | 'memoryContextLimit'
  | 'memoryReverseWeight'
  | 'src'
  | 'order'
  | 'useGaslight'
>

/**
 * This is al
 */
export const adapterSettings: {
  [key in keyof PresetAISettings]: AIAdapter[]
} = {
  temp: ['kobold', 'novel', 'ooba', 'horde', 'openai', 'scale', 'claude', 'goose'],
  maxTokens: AI_ADAPTERS.slice(),
  maxContextLength: AI_ADAPTERS.slice(),
  antiBond: ['openai', 'scale'],

  gaslight: ['openai', 'novel', 'scale', 'kobold', 'claude', 'ooba', 'goose', 'openrouter'],
  systemPrompt: ['openai', 'novel', 'scale', 'kobold', 'claude', 'ooba', 'goose', 'openrouter'],
  ignoreCharacterSystemPrompt: [
    'openai',
    'novel',
    'scale',
    'kobold',
    'claude',
    'ooba',
    'goose',
    'openrouter',
  ],
  ultimeJailbreak: ['openai', 'claude', 'kobold', 'scale', 'openrouter', 'novel'],
  prefill: ['claude'],
  ignoreCharacterUjb: ['openai', 'claude', 'kobold', 'openrouter'],

  topP: ['horde', 'kobold', 'claude', 'ooba', 'openai', 'novel'],
  repetitionPenalty: ['horde', 'novel', 'kobold', 'ooba'],
  repetitionPenaltyRange: ['horde', 'novel', 'kobold'],
  repetitionPenaltySlope: ['horde', 'novel', 'kobold'],
  tailFreeSampling: ['horde', 'novel', 'kobold'],
  topA: ['horde', 'novel', 'kobold'],
  topK: ['horde', 'novel', 'kobold', 'ooba', 'claude'],
  typicalP: ['horde', 'novel', 'kobold', 'ooba'],

  topG: ['novel'],
  mirostatLR: ['novel'],
  mirostatTau: ['novel'],
  cfgScale: ['novel'],
  cfgOppose: ['novel'],

  thirdPartyUrl: ['kobold', 'ooba'],
  thirdPartyFormat: ['kobold'],
  claudeModel: ['claude', 'kobold'],
  novelModel: ['novel'],
  oaiModel: ['openai', 'kobold'],
  frequencyPenalty: ['openai', 'kobold', 'novel'],
  presencePenalty: ['openai', 'kobold', 'novel'],
  streamResponse: ['openai', 'kobold', 'novel', 'claude', 'ooba'],
  openRouterModel: ['openrouter'],

  addBosToken: ['ooba'],
  banEosToken: ['ooba'],
  encoderRepitionPenalty: ['ooba'],
  penaltyAlpha: ['ooba'],

  replicateModelVersion: ['replicate'],
  replicateModelType: ['replicate'],

  skipSpecialTokens: ['ooba', 'kobold'],
}

export type RegisteredAdapter = {
  name: AIAdapter
  settings: AdapterSetting[]
  options: Array<keyof PresetAISettings>
}

export const settingLabels: { [key in keyof PresetAISettings]: string } = {
  temp: 'Temperature',
  maxTokens: 'Max Tokens (Response length)',
  repetitionPenalty: 'Repetition Penalty',
  repetitionPenaltyRange: 'Repetition Penality Range',
  repetitionPenaltySlope: 'Repetition Penalty Slope',
  tailFreeSampling: 'Tail Free Sampling',
  topA: 'Top A',
  topK: 'Top K',
  topP: 'Top P',
  typicalP: 'Typical P',
  addBosToken: 'Add BOS Token',
  antiBond: 'Anti-bond',
  banEosToken: 'Ban EOS Token',
  cfgOppose: 'CFG Opposing Prompt',
  cfgScale: 'CFG Scale',
  claudeModel: 'Claude Model',
  encoderRepitionPenalty: 'Encoder Repetition Penalty',
  frequencyPenalty: 'Frequency Penalty',
  gaslight: 'Prompt Template',
  ignoreCharacterSystemPrompt: 'Ignore Character System Prompt',
  ignoreCharacterUjb: 'Ignore Character Jailbreak',
  maxContextLength: 'Max Context Length',
  memoryChatEmbedLimit: 'Memory: Chat Embed Context Limit',
  memoryUserEmbedLimit: 'Memory: User-specific Embed Context Limit',
  novelModel: 'NovelAI Model',
  oaiModel: 'OpenAI Model',
  openRouterModel: 'OpenRouter Model',
  penaltyAlpha: 'Penalty Alpha',
  presencePenalty: 'Presence Penalty',
  replicateModelName: 'Replicate Model',
  replicateModelType: 'Replicate Model Type',
  replicateModelVersion: 'Replicate Model Version',
  skipSpecialTokens: 'Skip Special Tokens',
  streamResponse: 'Stream Response',
  systemPrompt: 'System (Instruction) Prompt',
  thirdPartyFormat: 'Third Party Format',
  thirdPartyUrl: 'Third Party URL',
  ultimeJailbreak: 'Jailbreak',
  prefill: 'Bot response prefilling',
  useTemplateParser: 'Use V2 Prompt Parser',
  topG: 'Top G',
  mirostatTau: 'Mirostat Tau',
  mirostatLR: 'Mirostat LR',
}

export const samplerOrders: { [key in AIAdapter]?: Array<keyof PresetAISettings> } = {
  kobold: ['topK', 'topA', 'topP', 'tailFreeSampling', 'typicalP', 'temp', 'repetitionPenalty'],
  novel: [
    'temp',
    'topK',
    'topP',
    'tailFreeSampling',
    'topA',
    'typicalP',
    'cfgScale',
    'topG',
    'mirostatTau',
  ],
}
