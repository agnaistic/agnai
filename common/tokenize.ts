export type TokenCounter = (text: string) => number

export type Encoder = {
  count: (text: string) => number
  encode: (text: string) => number[]
  decode: (tokens: number[]) => string
}

export type Tokenizer = {
  decode: (tokens: Int32Array) => string
  encode: (value: string) => Int32Array
}

export async function encode(text: string) {
  const encoder = await import('gpt-3-encoder').then((mod) => mod.encode)
  return encoder(text)
}

export async function tokenize(text: string) {
  const encoder = await getEncoder()
  return encoder(text)
}

export async function decode(tokens: number[]) {
  const decoder = await getDecoder()
  return decoder(tokens)
}

export async function getEncoder() {
  const encoder = await import('gpt-3-encoder').then((mod) => mod.encode)
  return (text: string) => encoder(text).length
}

export async function getDecoder() {
  const decode = await import('gpt-3-encoder').then((mod) => mod.decode)
  return (tokens: number[]) => decode(tokens)
}
