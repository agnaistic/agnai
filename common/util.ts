import { AppSchema } from './types/schema'

export function replace<T extends { _id: string }>(id: string, list: T[], item: Partial<T>) {
  return list.map((li) => (li._id === id ? { ...li, ...item } : li))
}

export function findOne<T extends { _id: string }>(id: string, list: T[]): T | void {
  for (const item of list) {
    if (item._id === id) return item
  }
}

export function toArray<T>(values?: T | T[]): T[] {
  if (values === undefined) return []
  if (Array.isArray(values)) return values
  return [values]
}

export function distinct<T extends { _id: string }>(values: T[]) {
  const set = new Set<string>()
  const next: T[] = []

  for (const item of values) {
    if (set.has(item._id)) continue
    next.push(item)
    set.add(item._id)
  }

  return next
}
export function wait(secs: number) {
  return new Promise((res) => setTimeout(res, secs * 1000))
}

export function waitMs(ms: number) {
  return new Promise((res) => setTimeout(res, ms))
}

export function isObject(val: unknown) {
  return Object.prototype.toString.call(val) === '[object Object]'
}

export function cycleArray<T>(arr: T[], shiftAmount: number) {
  const effectiveShift = shiftAmount % arr.length
  return [...arr.slice(effectiveShift), ...arr.slice(0, effectiveShift)]
}

export const dateStringFromUnix = (unixTimestamp: number): string =>
  new Date(unixTimestamp).toISOString()

export function uniqBy<T, U>(arr: T[], accessor: (el: T) => U) {
  return arr.flatMap((el, i) => (arr.slice(0, i).map(accessor).includes(accessor(el)) ? [] : [el]))
}

export function toDuration(valueSecs: number, full?: boolean) {
  if (valueSecs === 0) {
    if (full) return '0s'
    return '0 seconds'
  }
  const {
    duration: [days, hours, minutes, seconds],
  } = toRawDuration(valueSecs)

  if (full) {
    return [`${days}d`, `${hours}h`, `${minutes}m`, `${seconds}s`]
      .filter((time) => !time.startsWith('0'))
      .join(':')
  }

  if (days) {
    return `${days} ${days > 1 ? 'days' : 'day'}`
  }

  if (hours) {
    return `${hours} ${hours > 1 ? 'hours' : 'hour'}`
  }

  if (minutes) {
    return `${minutes} ${minutes > 1 ? 'mins' : 'min'}`
  }

  return `${seconds} ${seconds > 1 ? 'seconds' : 'second'}`
}

export function elapsedSince(date: string | Date, offsetMs: number = 0) {
  if (!date) return 'unknown'
  const diff = Date.now() - new Date(date).valueOf() + offsetMs
  const elapsed = Math.floor(diff / 1000)
  if (elapsed < 60) return 'a moment'
  return toDuration(Math.floor(elapsed))
}

const ONE_HOUR = 3600
const ONE_DAY = 86400

type Days = number
type Hours = number
type Minutes = number
type Seconds = number

type Duration = [Days, Hours, Minutes, Seconds]

function toRawDuration(valueSecs: number) {
  const days = Math.floor(valueSecs / ONE_DAY)
  const hours = Math.floor(valueSecs / ONE_HOUR) % 24
  const mins = Math.floor(valueSecs / 60) % 60
  const secs = Math.ceil(valueSecs % 60)

  return {
    duration: [days, hours, mins, secs] as Duration,
    seconds: valueSecs,
    text: valueSecs <= 0 ? 'now' : `${days}d:${hours}h:${mins}m:${secs}s`.replace('0d:', ''),
  }
}

/**
 * Remove leading tabs and whitespace on each line from template strings
 * Also 'trim' the text before returning
 */
export function neat(params: TemplateStringsArray, ...rest: string[]) {
  let str = ''
  for (let i = 0; i < params.length; i++) {
    str += params[i] + (rest[i] || '')
  }

  return str
    .split('\n')
    .map((line) => line.replace(/^[\t ]+/g, ''))
    .join('\n')
    .trim()
}

const end = `."'*!?)}]\``.split('')
export function trimSentence(text: string) {
  const last = end.reduce((last, char) => {
    const index = text.lastIndexOf(char)
    return index > last ? index : last
  }, -1)

  if (last === -1) return text
  return text.slice(0, last + 1)
}

export function slugify(str: string) {
  return str
    .toLowerCase()
    .trim()
    .replace(/[^\w\s-]/g, '')
    .replace(/[\s_-]+/g, '-')
    .replace(/^-+|-+$/g, '')
}

// https://stackoverflow.com/a/3561711
export function escapeRegex(string: string) {
  return string.replace(/[/\-\\^$*+?.()|[\]{}]/g, '\\$&')
}

export function getBotName(
  chat: AppSchema.Chat,
  msg: AppSchema.ChatMessage,
  chars: Record<string, AppSchema.Character>,
  replyAs: AppSchema.Character,
  main: AppSchema.Character,
  impersonate?: AppSchema.Character
) {
  if (!msg.characterId) return replyAs?.name || main.name
  if (msg.characterId.startsWith('temp-')) {
    const temp = chat.tempCharacters?.[msg.characterId]
    if (!temp) return main.name
    return temp.name
  }

  const char =
    msg.characterId && impersonate?._id === msg.characterId ? impersonate : chars[msg.characterId]

  if (!char) {
    return main.name
  }

  return char.name
}

export function eventGenerator<T = any>() {
  const queue: Array<Promise<any>> = []
  let done = false
  let signal = false

  const stream = (async function* () {
    while (!done) {
      await waitMs(1)
      const promise = queue.shift()
      if (!promise) {
        if (signal) {
          done = true
        }
        continue
      }

      const result = await promise
      yield result
    }
  })() as AsyncGenerator<T, T>

  return {
    stream,
    push: (value: T) => {
      if (done) return
      queue.push(Promise.resolve(value))
    },
    done: () => {
      if (done) return
      signal = true
    },
  }
}

/**
 * Will automatically clamp to `-max -> max` if `min` is not provided
 */
export function clamp(toClamp: number, max: number, min?: number) {
  return Math.max(Math.min(toClamp, max), min ?? -max)
}

export function now() {
  return new Date().toISOString()
}
