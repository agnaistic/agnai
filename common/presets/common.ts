import { neat } from '../util'

export const gaslights = {
  alpaca: neat`
  Below is an instruction that describes a task. Write a response that appropriately completes the request.
  
  Write {{char}}'s next reply in a fictional roleplay chat between {{#each bot}}{{.name}}, {{/each}}{{char}}.
  
  {{char}}'s Persona: {{personality}}
  
  This scenario of the conversation: {{scenario}}
  
  This is how {{char}} should talk: {{example_dialogue}}
  
  Then the roleplay chat between {{#each bot}}{{.name}}, {{/each}}{{char}} begins.
  
  {{#each msg}}{{#if .isbot}}### Response:\n{{.name}}: {{.msg}}{{/if}}{{#if .isuser}}### Instruction:\n{{.name}}: {{.msg}}{{/if}}
  {{/each}}
  {{ujb}}
  ### Response:
  {{post}}
  `,
}
