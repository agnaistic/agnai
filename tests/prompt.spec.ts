import './init'
import { expect } from 'chai'
import { OPENAI_MODELS } from '../common/adapters'
import { BOT_REPLACE, SELF_REPLACE } from '../common/prompt'
import { toChat, build, botMsg, toMsg, entities, reset } from './util'
import { getTokenCounter } from '../srv/tokenize'

const { chat, replyAs, main } = entities

describe('Prompt building', () => {
  before(reset)

  it('will build a basic prompt', () => {
    const actual = build([botMsg('FIRST'), toMsg('SECOND')])
    expect(actual.template).toMatchSnapshot()
  })

  it('will build a continue prompt', () => {
    const actual = build([botMsg('FIRST')], { continue: 'ORIGINAL', replyAs: main })
    expect(actual.template).toMatchSnapshot()
  })

  // i dont understand this test. @malfoyslastname 2023-06-07
  it('will exclude sample chat when gaslight contains sample chat placeholder', () => {
    const actual = build([botMsg('FIRST'), toMsg('SECOND')], {
      continue: 'ORIGINAL',
      chat: { ...chat, adapter: 'openai' },
      settings: {
        oaiModel: OPENAI_MODELS.Turbo,
        gaslight: `{{char}}'s Persona: {{personality}}
Scenario: {{scenario}}
This is how {{char}} should talk: {{example_dialogue}}`,
        useGaslight: true,
      },
    })
    expect(actual.template).toMatchSnapshot()
  })

  it('will include sample chat when gaslight does not contain sample chat placeholder', () => {
    const actual = build([botMsg('FIRST'), toMsg('SECOND')], {
      continue: 'ORIGINAL',
      chat,
      settings: { service: 'openai', oaiModel: OPENAI_MODELS.Turbo, gaslight: 'Gaslight' },
    })
    expect(actual.template).toMatchSnapshot()
  })

  it('will include will two memories by weight when triggered', () => {
    const actual = build([botMsg('FIRST'), toMsg('10-TRIGGER'), toMsg('1-TRIGGER')])
    expect(actual.template).toMatchSnapshot()
  })

  it('will exclude lowest priority memory to fit in budget', () => {
    const limit = getTokenCounter('main')(`ENTRY ONE. ENTRY TWO. ENTREE THREE.`) - 1
    const actual = build(
      [botMsg('FIRST'), toMsg('1-TRIGGER'), toMsg('10-TRIGGER'), toMsg('20-TRIGGER')],
      {
        settings: { memoryContextLimit: limit },
      }
    )
    expect(actual.template).toMatchSnapshot()
  })

  it('will order by trigger position when weight tie occurs', () => {
    const actual = build([
      botMsg('FIRST'),
      toMsg('1-TRIGGER'),
      toMsg('TIE-TRIGGER'),
      toMsg('20-TRIGGER'),
    ])
    expect(actual.template).toMatchSnapshot()
  })

  it('will exclude matches that are not a whole word match', () => {
    const actual = build([botMsg('LONGWOR A'), toMsg('1-TRIGGER')], {
      settings: { memoryDepth: 2 },
    })

    expect(actual.template).toMatchSnapshot()
  })

  it('will exclude memories triggered outside of memory depth', () => {
    const actual = build(
      [botMsg('FIRST'), toMsg('1-TRIGGER'), toMsg('TIE-TRIGGER'), toMsg('20-TRIGGER')],
      {
        settings: { memoryDepth: 2 },
      }
    )

    expect(actual.template).toMatchSnapshot()
  })

  it('will include gaslight for non-turbo adapter', () => {
    const actual = build(
      [botMsg('FIRST'), toMsg('1-TRIGGER'), toMsg('TIE-TRIGGER'), toMsg('20-TRIGGER')],
      {
        settings: { gaslight: 'GASLIGHT {{user}}', useGaslight: true },
      }
    )

    expect(actual.template).toMatchSnapshot()
  })

  it('will include placeholders in the gaslight', () => {
    const actual = build(
      [botMsg('FIRST'), toMsg('1-TRIGGER'), toMsg('TIE-TRIGGER'), toMsg('20-TRIGGER')],
      {
        settings: {
          gaslight: 'GASLIGHT\n{{user}}\n{{char}}\nFacts:{{memory}}',
          useGaslight: true,
        },
      }
    )

    expect(actual.template).toMatchSnapshot()
  })

  it('will not use the gaslight when set to false', () => {
    const actual = build(
      [botMsg('FIRST'), toMsg('1-TRIGGER'), toMsg('TIE-TRIGGER'), toMsg('20-TRIGGER')],
      {
        settings: {
          gaslight: 'GASLIGHT\n{{user}}\n{{char}}\nFacts: {{memory}}',
          useGaslight: false,
        },
      }
    )

    expect(actual.template).toMatchSnapshot()
  })

  it('uses the correct replaces for all instances of {{char}}, {{user}}, <BOT>, and <USER>, case insensitive', () => {
    const input =
      '{{char}} loves {{user}}, {{CHAR}} hates {{USER}}, {{Char}} eats {{User}}, <BOT> drinks <USER>, <bot> boops <user>, <Bot> kicks <User>'
    const expected =
      'Haruhi loves Chad, Haruhi hates Chad, Haruhi eats Chad, Haruhi drinks Chad, Haruhi boops Chad, Haruhi kicks Chad'
    const actual = input.replace(BOT_REPLACE, 'Haruhi').replace(SELF_REPLACE, 'Chad')
    expect(actual).to.equal(expected)
  })

  it('will use correct placeholders in scenario and sample chat', () => {
    const actual = build([], { replyAs, char: main, chat: toChat(main) })
    expect(actual.template).toMatchSnapshot()
  })

  it('will omit sample chat when replyAs has no samplechat', () => {
    const actual = build([], {
      replyAs: { ...replyAs, sampleChat: '' },
      char: main,
      chat: toChat(main),
    })
    expect(actual.template).toMatchSnapshot()
  })

  it('will include example dialogue with omitted from template', () => {
    const actual = build(
      [botMsg('FIRST'), toMsg('1-TRIGGER'), toMsg('TIE-TRIGGER'), toMsg('20-TRIGGER')],
      {
        char: { ...main, sampleChat: 'Bot: Example_Dialogue' },
        settings: {
          gaslight: 'GASLIGHT\n{{user}}\n{{char}}\nFacts: {{memory}}',
          useGaslight: false,
        },
      }
    )

    expect(actual.template).toMatchSnapshot()
  })

  it('will include ujb when omitted from gaslight', () => {
    const actual = build([botMsg('bot response')], {
      char: main,
      settings: {
        useGaslight: true,
        gaslight: `GASLIGHT TEMPLATE\n{{user}}\n{{char}}\nAuto-injected text:`,
        ultimeJailbreak: '!!UJB_PROMPT!!',
      },
    })
    expect(actual).to.matchSnapshot()
  })
})

// function expected(...lines: string[]) {
//   // Replace spaces with dots to help with interpreting test results
//   return lines.join('\n').replace(/ /g, '.')
// }
