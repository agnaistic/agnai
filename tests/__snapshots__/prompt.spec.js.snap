// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prompt building will build a basic prompt 1`] = `
"OtherBot's Persona: OtherBot replies a lot

Scenario: MAIN MainChar
How OtherBot speaks: SAMPLECHAT OtherBot

<START>
MainChar: FIRST
ChatOwner: SECOND
OtherBot:"
`;

exports[`Prompt building will build a continue prompt 1`] = `
"MainChar's Persona: MainCharacter talks a lot

Scenario: MAIN MainChar
How MainChar speaks: SAMPLECHAT MainChar

<START>
MainChar: FIRST
MainChar: ORIGINAL
MainChar:"
`;

exports[`Prompt building will exclude lowest priority memory to fit in budget 1`] = `
"OtherBot's Persona: OtherBot replies a lot

Scenario: MAIN MainChar
Facts:ENTRY TWO
ENTRY THREE
How OtherBot speaks: SAMPLECHAT OtherBot

<START>
MainChar: FIRST
ChatOwner: 1-TRIGGER
ChatOwner: 10-TRIGGER
ChatOwner: 20-TRIGGER
OtherBot:"
`;

exports[`Prompt building will exclude matches that are not a whole word match 1`] = `
"OtherBot's Persona: OtherBot replies a lot

Scenario: MAIN MainChar
Facts:ENTRY ONE
How OtherBot speaks: SAMPLECHAT OtherBot

<START>
MainChar: LONGWOR A
ChatOwner: 1-TRIGGER
OtherBot:"
`;

exports[`Prompt building will exclude memories triggered outside of memory depth 1`] = `
"OtherBot's Persona: OtherBot replies a lot

Scenario: MAIN MainChar
Facts:ENTRY TIE
ENTRY THREE
How OtherBot speaks: SAMPLECHAT OtherBot

<START>
MainChar: FIRST
ChatOwner: 1-TRIGGER
ChatOwner: TIE-TRIGGER
ChatOwner: 20-TRIGGER
OtherBot:"
`;

exports[`Prompt building will exclude sample chat when gaslight contains sample chat placeholder 1`] = `
"OtherBot's Persona: OtherBot replies a lot
Scenario: MAIN MainChar
This is how OtherBot should talk: SAMPLECHAT OtherBot
MainChar: FIRST
ChatOwner: SECOND
MainChar: ORIGINAL
OtherBot:"
`;

exports[`Prompt building will exclude scenario when empty in condition 1`] = `
"Roleplay instructions
How OtherBot speaks:
SAMPLECHAT OtherBot
<START>
MainChar: first
OtherBot:"
`;

exports[`Prompt building will include example dialogue with omitted from template (no longer true) 1`] = `
"GASLIGHT
ChatOwner
OtherBot
Facts: ENTRY ONE
ENTRY TIE
ENTRY THREE
How OtherBot speaks:
SAMPLECHAT OtherBot
<START>
MainChar: FIRST
ChatOwner: 1-TRIGGER
ChatOwner: TIE-TRIGGER
ChatOwner: 20-TRIGGER
OtherBot:"
`;

exports[`Prompt building will include example dialogue with omitted from template 1`] = `
"OtherBot's Persona: OtherBot replies a lot
Scenario: MAIN MainChar
Facts:ENTRY ONE
ENTRY TIE
ENTRY THREE

How OtherBot speaks: SAMPLECHAT OtherBot
<START>
MainChar: FIRST
ChatOwner: 1-TRIGGER
ChatOwner: TIE-TRIGGER
ChatOwner: 20-TRIGGER
OtherBot:"
`;

exports[`Prompt building will include gaslight for non-turbo adapter 1`] = `
"GASLIGHT ChatOwner
How OtherBot speaks:
SAMPLECHAT OtherBot
<START>
MainChar: FIRST
ChatOwner: 1-TRIGGER
ChatOwner: TIE-TRIGGER
ChatOwner: 20-TRIGGER
OtherBot:"
`;

exports[`Prompt building will include placeholders in the gaslight 1`] = `
"GASLIGHT
ChatOwner
OtherBot
Facts:ENTRY ONE
ENTRY TIE
ENTRY THREE
How OtherBot speaks:
SAMPLECHAT OtherBot
<START>
MainChar: FIRST
ChatOwner: 1-TRIGGER
ChatOwner: TIE-TRIGGER
ChatOwner: 20-TRIGGER
OtherBot:"
`;

exports[`Prompt building will include sample chat when gaslight does not contain sample chat placeholder 1`] = `
"Gaslight
How OtherBot speaks:
SAMPLECHAT OtherBot
System: New conversation started. Previous conversations are examples only.
MainChar: FIRST
ChatOwner: SECOND
MainChar: ORIGINAL
OtherBot:"
`;

exports[`Prompt building will include scenario when populated in condition 1`] = `
"Roleplay instructions
Scenario: Populated scenario
How OtherBot speaks:
SAMPLECHAT OtherBot
<START>
MainChar: first
OtherBot:"
`;

exports[`Prompt building will include ujb when omitted from gaslight 1`] = `
Object {
  "lines": Array [
    "MainChar: bot response",
  ],
  "parts": Object {
    "allPersonas": Array [
      "MainChar's personality: MainCharacter talks a lot",
    ],
    "chatEmbeds": Array [],
    "greeting": "GREETING",
    "memory": "",
    "persona": "OtherBot replies a lot",
    "post": Array [
      "OtherBot:",
    ],
    "sampleChat": Array [
      "SAMPLECHAT OtherBot",
    ],
    "scenario": "MAIN MainChar",
    "systemPrompt": "",
    "ujb": "",
    "userEmbeds": Array [],
  },
  "template": Object {
    "inserts": Map {},
    "parsed": "GASLIGHT TEMPLATE
ChatOwner
OtherBot
Auto-injected text:
How OtherBot speaks:
SAMPLECHAT OtherBot
<START>
MainChar: bot response
OtherBot:",
  },
}
`;

exports[`Prompt building will include will two memories by weight when triggered 1`] = `
"OtherBot's Persona: OtherBot replies a lot

Scenario: MAIN MainChar
Facts:ENTRY ONE
ENTRY TWO
How OtherBot speaks: SAMPLECHAT OtherBot

<START>
MainChar: FIRST
ChatOwner: 10-TRIGGER
ChatOwner: 1-TRIGGER
OtherBot:"
`;

exports[`Prompt building will not use the gaslight when set to false 1`] = `
"GASLIGHT
ChatOwner
OtherBot
Facts: ENTRY ONE
ENTRY TIE
ENTRY THREE
How OtherBot speaks:
SAMPLECHAT OtherBot
<START>
MainChar: FIRST
ChatOwner: 1-TRIGGER
ChatOwner: TIE-TRIGGER
ChatOwner: 20-TRIGGER
OtherBot:"
`;

exports[`Prompt building will omit sample chat when replyAs has no samplechat 1`] = `
"OtherBot's Persona: OtherBot replies a lot

Scenario: MAIN MainChar
<START>

OtherBot:"
`;

exports[`Prompt building will order by trigger position when weight tie occurs 1`] = `
"OtherBot's Persona: OtherBot replies a lot

Scenario: MAIN MainChar
Facts:ENTRY ONE
ENTRY TIE
ENTRY THREE
How OtherBot speaks: SAMPLECHAT OtherBot

<START>
MainChar: FIRST
ChatOwner: 1-TRIGGER
ChatOwner: TIE-TRIGGER
ChatOwner: 20-TRIGGER
OtherBot:"
`;

exports[`Prompt building will use correct placeholders in scenario and sample chat 1`] = `
"OtherBot's Persona: OtherBot replies a lot

Scenario: MAIN MainChar
How OtherBot speaks: SAMPLECHAT OtherBot

<START>

OtherBot:"
`;
